src/amd/compiler/aco_lower_bool_phis.cpp

```diff
@@ -206,35 +209,32 @@ void lower_linear_bool_phi(Program *program, Block *block, aco_ptr<Instruction>&
 void lower_bool_phis(Program* program)
 {
    for (Block& block : program->blocks) {
-      std::vector<aco_ptr<Instruction>> instructions;
-      std::vector<aco_ptr<Instruction>> non_phi;
-      instructions.swap(block.instructions);
-      block.instructions.reserve(instructions.size());
-      unsigned i = 0;
-      for (; i < instructions.size(); i++)
-      {
-         aco_ptr<Instruction>& phi = instructions[i];
-         if (phi->opcode != aco_opcode::p_phi && phi->opcode != aco_opcode::p_linear_phi)
-            break;
-         if (phi->opcode == aco_opcode::p_phi && phi->definitions[0].regClass() == s2) {
-            non_phi.emplace_back(std::move(lower_divergent_bool_phi(program, &block, phi)));
-         } else if (phi->opcode == aco_opcode::p_linear_phi && phi->definitions[0].regClass() == s1) {
+      // find the first non linear phi
+      auto IsNonlinearPhi = [] (const aco_ptr<Instruction>& phi) -> bool {
+         return phi->opcode != aco_opcode::p_phi && phi->opcode != aco_opcode::p_linear_phi;
+      };
+      auto end = std::find_if(block.instructions.begin(), block.instructions.end(), IsNonlinearPhi);
+
+      // partition divergent phi's to the back
+      auto IsNotDivergentPhi = [] (const aco_ptr<Instruction>& phi) -> bool {
+         return !(phi->opcode == aco_opcode::p_phi && phi->definitions[0].regClass() == s2);
+      };
+      auto middle = std::stable_partition(block.instructions.begin(), end, IsNotDivergentPhi);
+
+      // lower the phi's
+      for (auto it = block.instructions.begin(); it != middle; ++it) {
+         aco_ptr<Instruction>& phi = *it;
+         if (phi->opcode == aco_opcode::p_linear_phi && phi->definitions[0].regClass() == s1) {
             /* if it's a valid non-boolean phi, this should be a no-op */
             lower_linear_bool_phi(program, &block, phi);
-            block.instructions.emplace_back(std::move(phi));
-         } else {
-            block.instructions.emplace_back(std::move(phi));
          }
       }
-      for (auto&& instr : non_phi) {
-         assert(instr->opcode != aco_opcode::p_phi && instr->opcode != aco_opcode::p_linear_phi);
-         block.instructions.emplace_back(std::move(instr));
-      }
-      for (; i < instructions.size(); i++) {
-         aco_ptr<Instruction> instr = std::move(instructions[i]);
-         assert(instr->opcode != aco_opcode::p_phi && instr->opcode != aco_opcode::p_linear_phi);
-         block.instructions.emplace_back(std::move(instr));
+      for (auto it = middle; it != end; ++it) {
+         aco_ptr<Instruction>& phi = *it;
+         lower_divergent_bool_phi(program, &block, phi);
```

Comment:
This causes the result of lower_divergent_bool_phi() to be ignored, when it should actually be placed right after the block's phis.

Also, the instruction passed to `lower_divergent_bool_phi` should be removed and the `lower_divergent_bool_phi` call can end up inserting new instructions (and invalidating iterators).

Reply:
I cannot follow you here. `lower_divergent_bool_phi` takes a reference to the `aco_ptr` so the instruction will be rewritten inplace. Insofar there is no need to insert/remove.

The additional instructions are indeed troubling. I have to check later

Reply:
@pendingchaos aren't all additional instructions inserted at the predecessors and the phi replaced by a single s_mov (which is also the reason why these are moved down, so that the s_mov appears after the phis and not in the middle of them)?

Reply:
The phi isn't replaced by a single s_mov. A single s_mov is pushed to `non_phi` and the divergent boolean phi simply isn't pushed to the new instruction list. The function takes a `aco_ptr<Instruction>&` because `aco_ptr<T>` is a `unique_ptr` with a custom deleter and it didn't occur to me at the time to make it `const`.

With the partition thing, I think we can have `lower_divergent_bool_phi` return nothing and do this at the end instead:
```c++
phi.reset(bld.sop1(aco_opcode::s_mov_b64, phi->definitions[0], get_ssa(program, block->index, &state)).get_ptr());
```

Now sure yet how to solve the iterator invalidation issue. Maybe we can not ever create the s_mov and instead have `get_ssa` replace the boolean phi instead of creating a new one?

Reply:
I think it is better to just determine
```
const size_t endSize = std::distance(block.instructions.begin(), end);
```

And use a size_t based classical loop without iterators for the lowering.

I will do it tomorrow

Reply:
Nah did it anyway